src/main.rs

```diff
@@ -0,0 +1,253 @@
+
+use clap::{App, AppSettings, Arg, SubCommand};
+use manticore::io::StdWrite;
+use manticore::protocol;
+use manticore::protocol::firmware_version;
+use manticore::protocol::wire::FromWire;
+use manticore::protocol::wire::ToWire;
+use manticore::protocol::CommandType;
+use manticore::protocol::Header;
+use serde_json;
+use std::fs::OpenOptions;
+use std::io::prelude::*;
+use std::io::BufReader;
+
+/// Deserializes a message in JSON format from `reader` and then serializes the message in wire
+/// format to `writer`.
+fn from_json_to_wire<'a, T, R, W>(reader: R, writer: W)
+where
+    T: ToWire + serde::de::Deserialize<'a>,
+    R: std::io::Read,
+    W: manticore::io::Write,
+{
+    let mut de = serde_json::Deserializer::from_reader(reader);
+    let msg = T::deserialize(&mut de).expect("failed to deserialize JSON");
+    msg.to_wire(writer).expect("failed to write request");
+}
+
+/// Deserializes a message in JSON format from `input_file` and then serializes a header + message
+/// in wire format to.`output_file`. Uses `cmd_type` and `is_request` to determine the message type
+///  since the JSON format does not include the message header.
+fn from_json(
+    cmd_type: CommandType,
+    is_request: bool,
+    input_file: &str,
+    output_file: &str,
+) {
+    let mut output = OpenOptions::new()
+        .write(true)
+        .truncate(true)
+        .create(true)
+        .open(&output_file)
+        .expect("failed to open output file");
+
+    let input = OpenOptions::new()
+        .read(true)
+        .open(&input_file)
+        .expect("failed to open input file");
+    let input_reader = BufReader::new(input);
+
+    let mut stdwrite = StdWrite(&mut output);
+
+    let header = Header {
+        is_request: is_request,
+        command: cmd_type,
+    };
+    header
+        .to_wire(&mut stdwrite)
+        .expect("failed to write header");
+
+    if is_request {
+        match cmd_type {
+            CommandType::FirmwareVersion => {
+                from_json_to_wire::<
+                    firmware_version::FirmwareVersionRequest,
+                    _,
+                    _,
+                >(input_reader, stdwrite)
+            }
+            _ => panic!("Unsupported message type"),
+        }
+    } else {
+        match cmd_type {
+            CommandType::FirmwareVersion => {
+                from_json_to_wire::<
+                    firmware_version::FirmwareVersionResponse,
+                    _,
+                    _,
+                >(input_reader, stdwrite)
+            }
+            _ => panic!("Unsupported message type"),
+        }
+    }
+}
+
+/// Macro to deserialize wire format from an input file and then run an operation on the
+/// deserialized message.
+/// Arguments:
+/// * `input_file`: Identifier to get the input file name from.
+/// * `msg`: Identifier to hold the message to operator.
+/// * `op`: Expression to execute after deserializing `msg` from `input_file`.
+macro_rules! read_wire_and_operate {
```

Comment:
The inside of a macro won't be formatter by the rust fomatter (because reasons). Would you mind wrapping all these lines to 80 characters?

However, I believe you can use [erased_serde](https://docs.rs/erased-serde/0.3.12/erased_serde/) to turn this into a function.

Reply:
Oh yes, I did not know that. I'll look into `erased_serde` in a separate PR.